ortools/sat/csharp/CpModel.cs

```diff
@@ -99,37 +95,51 @@ public Constraint AddLinearExpressionInDomain(LinearExpr linear_expr, Domain dom
             linear.Vars.Add(term.Key.Index);
             linear.Coeffs.Add(term.Value);
         }
-        foreach (long value in domain.FlattenedIntervals())
+
+        linear.Domain.Add(lb is Int64.MinValue or Int64.MaxValue ? lb : lb - constant);
+        linear.Domain.Add(ub is Int64.MinValue or Int64.MaxValue ? ub : ub - constant);
+
+        ct.Proto.Linear = linear;
+        return ct;
+    }
+
+    public Constraint AddLinearExpressionInDomain(LinearExpr linear_expr, long[] flattenedIntervals)
```

Comment:
No. I do not want to show this API with flattened intervals. It is too complex. I must only be accessible in the Domain class.

Reply:
Is it acceptable if we switch it to private instead of public? So the library code could use it since it should know how to do it right?

Reply:
Domain can have more that 2 bounds. So I need a code that loops over all values.

Reply:
There are two use cases:
  linear_expr in [lb , ub] -> we have an API for that and we can reuse your code for it
  linear_expr in domain. -> no need to add an intermediate method with flattened intervals

The only common use case of the second option is expr != value. We could have a private method that does the right thing (add int_min, value - offset - 1, value - offset + 1, int_max) with little checks

Reply:
Doesn't it do that in line 117-120?

Reply:
I have pushed the relevant code on master.

Reply:
Nice change, that solves it very elegantly!
I'll update the PR to incorporate it.

Reply:
I have tweaked it:
  - create a private method to share common code
  - changed public to private on the NotEqualCst method